src/wp-includes/webfonts-api/providers/class-wp-webfonts-provider.php

```diff
@@ -0,0 +1,216 @@
+<?php
+/**
+ * Webfonts API: Provider abstract class.
+ *
+ * Individual webfonts providers should extend this class and implement.
+ *
+ * @package    WordPress
+ * @subpackage WebFonts
+ * @since      5.9.0
+ */
+
+/**
+ * Abstract class for Webfonts API providers.
+ *
+ * @since 5.9.0
+ */
+abstract class WP_Webfonts_Provider {
+
+	/**
+	 * The provider's unique ID.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var string
+	 */
+	protected $id;
+
+	/**
+	 * The provider's root URL.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var string
+	 */
+	protected $root_url = '';
+
+	/**
+	 * Webfonts to be processed.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var array[]
+	 */
+	protected $webfonts = array();
+
+	/**
+	 * Array of resources hints.
+	 *
+	 * Keyed by relation-type:
+	 *
+	 *      @type string $key => @type array resource hint.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var array
+	 */
+	protected $resource_hints = array();
+
+	/**
+	 * Whether the provider fetches external resources or not.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var bool
+	 */
+	protected $is_external = true;
+
+	/**
+	 * Get the provider's unique ID.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string
+	 */
+	public function get_id() {
+		return $this->id;
+	}
+
+	/**
+	 * Get the root URL for the provider.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string
+	 */
+	public function get_root_url() {
+		return $this->root_url;
+	}
+
+	/**
+	 * Sets the webfonts.
+	 *
+	 * The webfonts have been validated, are in kebab_case, and
+	 * are arranged by provider.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array[] $webfonts Registered webfonts.
+	 */
+	public function set_webfonts( array $webfonts ) {
+		$this->webfonts = $webfonts;
+
+		foreach ( $this->webfonts as $registered_key => $webfont ) {
+			$this->webfonts[ $registered_key ] = $this->prepare( $webfont );
+		}
+	}
+
+	/**
+	 * Prepares the given webfont.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array $webfont Webfont to validate.
+	 * @return array
+	 */
+	protected function prepare( array $webfont ) {
+		return $webfont;
+	}
```

Comment:
I raised this originally in https://github.com/WordPress/wordpress-develop/pull/1736#discussion_r736984593, and that part of my comment hasn't been addressed: Do we really need a separate `prepare` method? This feels a bit over-engineered to me, for the following reasons:

* Whatever can be done within this method, could also be implemented within the `get_css()` method.
* This even has potential performance implications: Why should whatever logic is here be run on registration rather than when getting the CSS? Running it on registration will run the logic in many cases where it isn't even needed. Moving the logic to `get_css()` and removing `prepare()` entirely will ensure the logic is only run when it's actually needed.
* This method is only used by the "local" provider and is otherwise just a no-op - another pointer that it seems overkill to include it as a separate method in the base class.